src/unix/filesystem.h

```diff
@@ -52,6 +52,35 @@ static inline const char *filename_from_path(const char *path)
     return filename;
 }
 
+/* Expects an empty buffer, and never resizes the buffer. */
+static inline void dirname_from_path(buffer dest, const char *path)
+{
+    int pathlen = runtime_strlen(path);
+    const char *last_delim = (char *)utf8_find_r((u8 *)path, '/');
+    const char *dirname;
+    int len;
+    if (!last_delim) {
+        dirname = path;
+        len = pathlen;
+    } else if (last_delim < path + pathlen - 1) {
+        dirname = last_delim + 1;
+        len = pathlen - (last_delim + 1 - path);
+    } else {    /* The path ends with '/'. */
+        const char *delim = (char *)utf8_findn_r((u8 *)path, last_delim - path,
+            '/');
+        if (!delim) {
+            dirname = path;
+            len = pathlen - 1;
+        } else {
+            dirname = delim + 1;
+            len = last_delim - (delim + 1);
+        }
+    }
+    if (len < dest->length)
```

Comment:
Nit # 3: Maybe just return false if the dirname won't fit instead of doing the peek_char?

Reply:
Yes, it looks better.